### PR TITLE
Nuxt support

### PR DESCRIPTION
**Pull request by cfofiu**
### Description

This PR handles basic Nuxt 2 and Next 3 support.

### Scenarios Tested

Covers basic Nuxt 3 with limited configuration.
Nuxt 2 has a couple more use cases, but the following should be handled:
* `ssr: false`
* `ssr: true`
* `target: 'server'`
* `target: 'static'`
* custom generate directory

### Sample Commands
Deploying Nuxt 2 and 3 codebase via `firebase emulators:start`.

### CODE REVIEW

File: src/frameworks/nuxt/index.ts [hunk 1]
```diff
@@ -1,33 +1,39 @@
 import { copy, pathExists } from "fs-extra";
 import { readFile } from "fs/promises";
-import { basename, join } from "path";
+import { join } from "path";
 import { gte } from "semver";
-import { BuildResult, findDependency, FrameworkType, relativeRequire, SupportLevel } from "..";
+import { findDependency, FrameworkType, relativeRequire, SupportLevel } from "..";
 
-export const name = "Nuxt";
+export const name = "Nuxt 3";
```
Comment (jamesdaniels) [inline src/frameworks/nuxt/index.ts:7]: I think we can drop the 3 here.

File: src/frameworks/nuxt/index.ts [hunk 2]
```diff
@@ -1,33 +1,39 @@
...
   const configFilesExist = await Promise.all([
     pathExists(join(dir, "nuxt.config.js")),
     pathExists(join(dir, "nuxt.config.ts")),
   ]);
+
   const anyConfigFileExists = configFilesExist.some((it) => it);
   if (!anyConfigFileExists && !nuxtDependency) return;
-  return { mayWantBackend: true };
+  if (!version) throw new Error("Unable to find the nuxt dep.");
+  if (gte(version, "3.0.0")) return { mayWantBackend: true };
```
Comment (jamesdaniels) [inline src/frameworks/nuxt/index.ts:23]: Let's keep it gte 3.0.0-0 so we capture all the RCs

File: src/frameworks/nuxt/index.ts [hunk 3]
```diff
@@ -1,33 +1,39 @@
...
+  const version = nuxtDependency?.version;
   const configFilesExist = await Promise.all([
     pathExists(join(dir, "nuxt.config.js")),
     pathExists(join(dir, "nuxt.config.ts")),
   ]);
+
   const anyConfigFileExists = configFilesExist.some((it) => it);
   if (!anyConfigFileExists && !nuxtDependency) return;
-  return { mayWantBackend: true };
+  if (!version) throw new Error("Unable to find the nuxt dep.");
```
Comment (jamesdaniels) [inline src/frameworks/nuxt/index.ts:22]: IMO we can drop these, let it fall back to the can't find framework error

File: src/frameworks/nuxt2/index.ts [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+}
+
+/**
+ * Get the Nuxt app
+ * @param cwd
+ * @return Nuxt app object
+ */
+async function getNuxtAppForBuild(cwd: string) {
+  let nuxt: any = null;
+  try {
```
Comment (jamesdaniels) [inline src/frameworks/nuxt2/index.ts:63]: IMO we should be able to drop the try/catch

File: src/frameworks/nuxt2/index.ts [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+  const version = nuxtDependency?.version;
+  const configFilesExist = await Promise.all([
+    pathExists(join(dir, "nuxt.config.js")),
+    pathExists(join(dir, "nuxt.config.ts")),
+  ]);
+  const anyConfigFileExists = configFilesExist.some((it) => it);
+
+  if (!anyConfigFileExists && !nuxtDependency) return;
+  if (!version) throw new Error("Unable to find the nuxt dep.");
+  if (lt(version, "3.0.0")) return { mayWantBackend: true };
```
Comment (jamesdaniels) [inline src/frameworks/nuxt2/index.ts:23]: 3.0.0-0 so it doesn't trip on RCs

File: src/frameworks/nuxt2/index.ts [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+  const nuxtDependency = findDependency("nuxt", { cwd: dir, depth: 0, omitDev: false });
+  const version = nuxtDependency?.version;
+  const configFilesExist = await Promise.all([
+    pathExists(join(dir, "nuxt.config.js")),
+    pathExists(join(dir, "nuxt.config.ts")),
+  ]);
+  const anyConfigFileExists = configFilesExist.some((it) => it);
+
+  if (!anyConfigFileExists && !nuxtDependency) return;
+  if (!version) throw new Error("Unable to find the nuxt dep.");
```
Comment (jamesdaniels) [inline src/frameworks/nuxt2/index.ts:22]: drop

File: src/frameworks/nuxt2/index.ts [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+		for the `.nuxt` directory in the destination directory.
+	*/
+  await copy(join(sourceDir, ".nuxt"), join(destDir, ".nuxt"));
+
+  /*
+		When using `SSR: false`, we need to copy the `nuxt.config.js` to the destination directory (`functions`)
+		This is because `loadNuxt` (called from `firebase-frameworks`) will look
+		for the `nuxt.config.js` file in the destination directory.
+		*/
+  if (nuxtConfig.ssr === false) {
```
Comment (jamesdaniels) [inline src/frameworks/nuxt2/index.ts:124]: nit, `!nuxtConfig.ssr`... that said if ssr is false should we skip entirely—by returning wantsBackend: false in build?

File: src/frameworks/nuxt2/index.ts [hunk 8]
```diff
@@ -0,0 +1,130 @@
+import { copy, pathExists } from "fs-extra";
+import { readFile } from "fs/promises";
+import { join } from "path";
+import { lt } from "semver";
+import { findDependency, FrameworkType, relativeRequire, SupportLevel } from "..";
+
+export const name = "Nuxt 2";
```
Comment (jamesdaniels) [inline src/frameworks/nuxt2/index.ts:7]: Can just call nuxt IMO

File: src/frameworks/nuxt2/index.ts [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+export const support = SupportLevel.Experimental;
+export const type = FrameworkType.Toolchain;
```
Comment (jamesdaniels) [inline src/frameworks/nuxt2/index.ts:9]: Metaframework

File: src/frameworks/nuxt/index.ts [hunk 10]
```diff
@@ -1,33 +1,39 @@
...
-import { basename, join } from "path";
+import { join } from "path";
 import { gte } from "semver";
-import { BuildResult, findDependency, FrameworkType, relativeRequire, SupportLevel } from "..";
+import { findDependency, FrameworkType, relativeRequire, SupportLevel } from "..";
 
-export const name = "Nuxt";
+export const name = "Nuxt 3";
 export const support = SupportLevel.Experimental;
 export const type = FrameworkType.Toolchain;
```
Comment (jamesdaniels) [inline src/frameworks/nuxt/index.ts:9]: huh, why was this toolchain, should be metaframework

Comment (cfofiu) [reply src/frameworks/nuxt/index.ts:9]: Well, [Toolchain](https://github.com/firebase/firebase-tools/blob/4e013b40a3d4069c3758235544d62d12e8c003d3/src/frameworks/index.ts#L89) seems to map to Vite?

File: src/frameworks/nuxt2/index.ts (hunk 7 above)
Comment (cfofiu) [reply src/frameworks/nuxt2/index.ts:124]: We still need a backend for `loadNuxt` to run

File: src/frameworks/nuxt2/index.ts (hunk 4 above)
Comment (cfofiu) [reply src/frameworks/nuxt2/index.ts]: implemented a small refactor; removed the try/catch block

File: src/frameworks/nuxt2/index.ts (hunk 7 above)
Comment (jamesdaniels) [reply src/frameworks/nuxt2/index.ts]: if you return `wantsBackend: false` in build it won't try to bundle up a functions directory at all

File: src/frameworks/nuxt/index.ts [hunk 11]
```diff
@@ -1,39 +1,53 @@
...
-  ]);
-  const anyConfigFileExists = configFilesExist.some((it) => it);
+  const nuxtDependency = findDependency("nuxt", {
+    cwd: dir,
+    depth: 0,
+    omitDev: false,
+  }) as NuxtDependency;
+
+  const version = nuxtDependency?.version;
+  const anyConfigFileExists = await nuxtConfigFilesExist(dir);
```
Comment (jamesdaniels) [inline src/frameworks/nuxt/index.ts:31]: nit, unused?

File: src/frameworks/nuxt2/index.ts [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+/**
+ * Copy the static files to the destination directory whether it's a static build or server build.
+ * @param root
+ * @param dest
+ */
+export async function ɵcodegenPublicDirectory(root: string, dest: string) {
+  const nuxt = await getNuxtApp(root);
+  const nuxtConfig = await nuxt.loadNuxtConfig();
+  const { ssr, target } = nuxtConfig;
+  /*
```
Comment (jamesdaniels) [inline src/frameworks/nuxt2/index.ts:105]: nit: spacing

**Review by jamesdaniels — APPROVED** [review]
LGTM better than what we have. Add changelog and can you fix Austin's email in a force-push?

File: src/frameworks/nuxt2/index.ts (hunk 12 above)
Comment (cfofiu) [reply src/frameworks/nuxt2/index.ts:105]: Comments formatting updated

**Pull request merged**